<a href="https://colab.research.google.com/github/JuanPedroDelCastillo/FisicaComputacional1/blob/master/Actividad3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Actividad 3**

Primero debemos **ejecutar la siguiente celda** para importar las bibliotecas utilizadas en los programas.

In [1]:
import numpy as np
import pandas as pd

## Para importar el archivo y guardarlo en una variable:

In [38]:
#Para importar el archivo csv tenemos que poner el link de github, en el cual se tiene que abrir en forma raw
url = "https://raw.githubusercontent.com/JuanPedroDelCastillo/FisicaComputacional1/master/Actividad3/26139.txt"

#Hacemos una lista de los encabezados
my_cols = ['Fecha', 'Precip', 'Evap', 'Tmax', 'Tmin']

#Creamos la lista con el comando pd.read_cvs()
#encoding 'cp1251' para poder usar símbolos en español
#sep = '\+' sirve para separar los datos con espacio
#skiprows sirve para saltarse renglones, hasta nuestra dataframe de interés
#skipfooter es para saltar renglones del final
#engine python para manipularlo en python
df_dist = pd.read_csv(url, names=my_cols, encoding='cp1251', sep='\s+', header=None, skiprows=19, skipfooter=1, engine='python')

## Para Imprimir el Inicio y Final:

In [39]:
#Imprime los primeros 5 renglones
print(df_dist.head())

#Imprime los últimos 5 renglones
print(df_dist.tail())

        Fecha Precip Evap  Tmax Tmin
0  01/01/1961      0  1.6    23    7
1  02/01/1961      0  2.1    25    5
2  03/01/1961      0  4.5    26    7
3  04/01/1961      0  2.5  26.5    8
4  05/01/1961     20  4.6    27  7.5
            Fecha Precip  Evap  Tmax  Tmin
19296  25/02/2018      0  Nulo  Nulo  Nulo
19297  26/02/2018      0  Nulo  Nulo  Nulo
19298  27/02/2018      0  Nulo  Nulo  Nulo
19299  28/02/2018      0  Nulo  Nulo  Nulo
19300  01/03/2018      0  Nulo  Nulo  Nulo


## Comandos para un Dataframe:

### Imprime el encabezado y final del dataframe:

In [40]:
#Imprimir las primeras n = 10 filas:
print(df_dist.head(10))

#Imprimir las últimas n = 10 filas:
print(df_dist.tail(10))

        Fecha Precip Evap  Tmax Tmin
0  01/01/1961      0  1.6    23    7
1  02/01/1961      0  2.1    25    5
2  03/01/1961      0  4.5    26    7
3  04/01/1961      0  2.5  26.5    8
4  05/01/1961     20  4.6    27  7.5
5  06/01/1961      0  3.2  27.5   10
6  07/01/1961      0  3.8    23   11
7  08/01/1961      0  3.1    24    7
8  09/01/1961      0  4.6    27  7.5
9  10/01/1961      0  2.5    28   10
            Fecha Precip  Evap  Tmax  Tmin
19291  20/02/2018      0  Nulo  Nulo  Nulo
19292  21/02/2018      0  Nulo  Nulo  Nulo
19293  22/02/2018      0  Nulo  Nulo  Nulo
19294  23/02/2018      0  Nulo  Nulo  Nulo
19295  24/02/2018      0  Nulo  Nulo  Nulo
19296  25/02/2018      0  Nulo  Nulo  Nulo
19297  26/02/2018      0  Nulo  Nulo  Nulo
19298  27/02/2018      0  Nulo  Nulo  Nulo
19299  28/02/2018      0  Nulo  Nulo  Nulo
19300  01/03/2018      0  Nulo  Nulo  Nulo


### ¿Qué dimensiones tiene tu dataframe?

In [41]:
#Para conocer el tamaño de la dataframe
df_dist.shape

(19301, 5)

### ¿Cómo es el contenido de tu dataframe?

In [42]:
df_dist.info

<bound method DataFrame.info of             Fecha Precip  Evap  Tmax  Tmin
0      01/01/1961      0   1.6    23     7
1      02/01/1961      0   2.1    25     5
2      03/01/1961      0   4.5    26     7
3      04/01/1961      0   2.5  26.5     8
4      05/01/1961     20   4.6    27   7.5
...           ...    ...   ...   ...   ...
19296  25/02/2018      0  Nulo  Nulo  Nulo
19297  26/02/2018      0  Nulo  Nulo  Nulo
19298  27/02/2018      0  Nulo  Nulo  Nulo
19299  28/02/2018      0  Nulo  Nulo  Nulo
19300  01/03/2018      0  Nulo  Nulo  Nulo

[19301 rows x 5 columns]>

### Reemplazar datos **Nulos**

In [43]:
#Primero realizamos una copia de trabajo y preservamos el dataframe original
df_work0 = df_dist.copy()
print(df_work0.head(3),
      '\n',
      df_work0.tail(3)
)

        Fecha Precip Evap Tmax Tmin
0  01/01/1961      0  1.6   23    7
1  02/01/1961      0  2.1   25    5
2  03/01/1961      0  4.5   26    7 
             Fecha Precip  Evap  Tmax  Tmin
19298  27/02/2018      0  Nulo  Nulo  Nulo
19299  28/02/2018      0  Nulo  Nulo  Nulo
19300  01/03/2018      0  Nulo  Nulo  Nulo


In [44]:
#Reemplazamos los textos 'Nulo' por espacio en blanco ''
str_Nulo = 'Nulo'
df_work1 = df_work0.replace(to_replace=str_Nulo, value='', regex=True)
print(df_work1.head(3),
      '\n',
      df_work1.tail(3)
)

        Fecha Precip Evap Tmax Tmin
0  01/01/1961      0  1.6   23    7
1  02/01/1961      0  2.1   25    5
2  03/01/1961      0  4.5   26    7 
             Fecha Precip Evap Tmax Tmin
19298  27/02/2018      0               
19299  28/02/2018      0               
19300  01/03/2018      0               


In [45]:
#Imprimimos la información actualizada del dataframe
df_work1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19301 entries, 0 to 19300
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Fecha   19301 non-null  object
 1   Precip  19301 non-null  object
 2   Evap    19301 non-null  object
 3   Tmax    19301 non-null  object
 4   Tmin    19301 non-null  object
dtypes: object(5)
memory usage: 754.1+ KB


### Convertir objetos a **Números Flotantes**

Para convertir los objetos a variables reales o flotantes, usamos for de tal modo que cada columna sustituya sus objetos a flotantes. Posteriormente se enseña la información modificada.

Además, se puede saber el número de datos nulos o vacíos con la función **df.isnull().sum()**.

In [46]:
#Usando el comando pd.to_numeric convertimos los objetos a variables reales
cols_list = ['Precip', 'Evap', 'Tmax', 'Tmin']
for cols in cols_list:
  df_work1[cols] = pd.to_numeric(df_work1[cols], errors='coerce')

#Se verán los datos ahora como flotantes
print(df_work1.info())

#Se muestran cuantos datos son nulos o vacíos
print(df_work1.isnull().sum())

#Se muestran los datos iniciales y finales
print(df_work1.head(3),
      '\n',
      df_work1.tail(3)
)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19301 entries, 0 to 19300
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Fecha   19301 non-null  object 
 1   Precip  19291 non-null  float64
 2   Evap    10285 non-null  float64
 3   Tmax    19121 non-null  float64
 4   Tmin    19138 non-null  float64
dtypes: float64(4), object(1)
memory usage: 754.1+ KB
None
Fecha        0
Precip      10
Evap      9016
Tmax       180
Tmin       163
dtype: int64
        Fecha  Precip  Evap  Tmax  Tmin
0  01/01/1961     0.0   1.6  23.0   7.0
1  02/01/1961     0.0   2.1  25.0   5.0
2  03/01/1961     0.0   4.5  26.0   7.0 
             Fecha  Precip  Evap  Tmax  Tmin
19298  27/02/2018     0.0   NaN   NaN   NaN
19299  28/02/2018     0.0   NaN   NaN   NaN
19300  01/03/2018     0.0   NaN   NaN   NaN


### Estadística de las variables numéricas

In [48]:
#A partir del comando df.describe() podemos obtener datos básicos de estadística
df_work1.describe().round(2)

Precip      Evap      Tmax      Tmin
count  19291.00  10285.00  19121.00  19138.00
mean       0.98      7.17     32.30     17.55
std        5.49      3.34      6.77      6.65
min        0.00      0.10      8.00     -3.00
25%        0.00      4.50     27.00     12.00
50%        0.00      6.90     33.00     17.00
75%        0.00      9.60     38.00     24.00
max      152.60     18.00     48.50     31.00

De estos datos se pueden analizar varios aspectos interesantes. Por ejemplo, la temperatura máxima en Hermosillo hasta la fecha de la recolepción de datos fue de 48.50 grados, mientras que la menor temperatura mínima ha sido -3.00 grados.
También se pueden analizar aspectos como los promedios y desviación estandar, por ejemplo el promedio de precipitación es de aproximadamente 1 mm y la desviación estandar de 5.49.  

Esta información también nos verifica si la información ha sido manipulada correctamente. Si se encuentran datos sin sentido puede advertirnos de posibles errores. En este caso se ve todo en orden.


### Convertir objeto a variable de **Tiempo**

In [49]:
#Antes de comenzar con nuevos pasos, es importante realizar una copia de
#seguridad para nuestra información
df_work2 = df_work1.copy()
df_work2['Fecha'].head()

0    01/01/1961
1    02/01/1961
2    03/01/1961
3    04/01/1961
4    05/01/1961
Name: Fecha, dtype: object

El tipo de variable **datetime64[ns]** contiene el año, mes y día de alguna fecha, para utilzarla convertimos de objeto a datetime. Para verificar el tipo de variable de las diferentes columnas usamos **dtypes**.

In [51]:
#Usamos dayfirst para que utilice el formato español de las fechas
df_work2['Fecha']=pd.to_datetime(df_work2['Fecha'], dayfirst=True).copy()

#dtypes para imprimir los tipos de variables
print(df_work2.dtypes)

Fecha     datetime64[ns]
Precip           float64
Evap             float64
Tmax             float64
Tmin             float64
dtype: object


Para añadir dos columnas extra en la que mostraremos el año y el mes utilizamos el siguiente comando:

In [54]:
#dt.year hace referencia al año de una variable datetime64.
df_work2['Año'] = df_work2['Fecha'].dt.year

#dt.month hace referencia al mes de una variable datetime64.
df_work2['Mes'] = df_work2['Fecha'].dt.month

#Verificamos que todo este en orden
print(df_work2.head(3),
      '\n',
      df_work2.tail(3),
      '\n',
      df_work2.dtypes
)



       Fecha  Precip  Evap  Tmax  Tmin   Año  Mes
0 1961-01-01     0.0   1.6  23.0   7.0  1961    1
1 1961-01-02     0.0   2.1  25.0   5.0  1961    1
2 1961-01-03     0.0   4.5  26.0   7.0  1961    1 
            Fecha  Precip  Evap  Tmax  Tmin   Año  Mes
19298 2018-02-27     0.0   NaN   NaN   NaN  2018    2
19299 2018-02-28     0.0   NaN   NaN   NaN  2018    2
19300 2018-03-01     0.0   NaN   NaN   NaN  2018    3 
 Fecha     datetime64[ns]
Precip           float64
Evap             float64
Tmax             float64
Tmin             float64
Año                int64
Mes                int64
dtype: object
